File: 279.perfect-squares.js
/*
 * @lc app=leetcode id=279 lang=javascript
 *
 * [279] Perfect Squares
 */

// @lc code=start
/**
 * @param {number} n
 * @return {number}
 */
var numSquares = function (n) {
  const dp = Array(n + 1).fill(Infinity);
  dp[0] = 0;
  for (let i = 0; i < n + 1; ++i) {
    for (let j = 1; i + j * j < n + 1; ++j) {
      dp[i + j * j] = Math.min(dp[i + j * j], dp[i] + 1);
    }
  }
  return dp[n];
};
// @lc code=end

File: 279.perfect-squares.py
#
# @lc app=leetcode id=279 lang=python3
#
# [279] Perfect Squares
#

# Time:  O(n * sqrt(n))
# Space: O(n)

# @lc code=start
class Solution:
    def numSquares(self, n: int) -> int:
        dp = [0]
        while len(dp) <= n:
            dp.append(min(dp[-i * i] for i in range(1, int(len(dp) ** 0.5 + 1))) + 1)
        return dp[n]


# @lc code=end
